test/hierarchical/test_ordinal.py

```diff
@@ -0,0 +1,103 @@
+from pathlib import Path
+from typing import Dict, List
+
+import numpy as np
+import petab
+import pytest
+
+import pypesto
+import pypesto.logging
+import pypesto.optimize
+import pypesto.petab
+
+inner_solver_options = [
+    [
+        {
+            'method': 'standard',
+            'reparameterized': False,
+            'intervalConstraints': 'max',
+        },
+        {
+            'method': 'reduced',
+            'reparameterized': False,
+            'intervalConstraints': 'max',
+        },
+        {
+            'method': 'reduced',
+            'reparameterized': True,
+            'intervalConstraints': 'max',
+        },
+    ],
+    [
+        {
+            'method': 'standard',
+            'reparameterized': False,
+            'intervalConstraints': 'max-min',
+        },
+        {
+            'method': 'reduced',
+            'reparameterized': False,
+            'intervalConstraints': 'max-min',
+        },
+        {
+            'method': 'reduced',
+            'reparameterized': True,
+            'intervalConstraints': 'max-min',
+        },
+    ],
+]
+
+example_ordinal_yaml = (
+    Path(__file__).parent
+    / '..'
+    / '..'
+    / 'doc'
+    / 'example'
+    / 'example_ordinal'
+    / 'example_ordinal.yaml'
+)
+
+
+@pytest.fixture(params=inner_solver_options)
+def inner_solver_options(request):
+    return request.param
+
+
+def test_evaluate_objective(inner_solver_options: List[Dict]):
+    """Check that standard / reduced / reparameterized formulations yield the
+    same result."""
+    petab_problem = petab.Problem.from_yaml(example_ordinal_yaml)
+    vals = []
+    for idx, option in enumerate(inner_solver_options):
+        problem = create_problem(petab_problem, option)
+        val = problem.objective(np.array([0, 0]))
+        vals.append(val)
+        assert np.isclose(vals[idx], vals[idx - 1])
+
+
+def test_optimization(inner_solver_options: List[Dict]):
+    """Check that optimizations finishes without error."""
+    petab_problem = petab.Problem.from_yaml(example_ordinal_yaml)
+
+    optimizer = pypesto.optimize.ScipyOptimizer(
+        method='Nelder-Mead', options={'maxiter': 10}
+    )
+    for option in inner_solver_options:
+        problem = create_problem(petab_problem, option)
+        pypesto.optimize.minimize(
+            problem=problem, n_starts=1, optimizer=optimizer
+        )
+
+
+def create_problem(
+    petab_problem: petab.Problem, option: Dict
+) -> pypesto.Problem:
+    importer = pypesto.petab.PetabImporter(petab_problem, ordinal=True)
+    importer.create_model()
+
+    objective = importer.create_objective(
+        inner_problem_method=option['method'],
+        inner_solver_options=option,
+    )
+    problem = importer.create_problem(objective)
+    return problem
```

Comment:
I think that given the complexity of the method, a substantially increased set of actual unit tests (i.e. test single components - functions, classes - instead of the full workflow) would be much more suited to ensuring correctness of the approach - e.g. regarding the solution of the inner problem, the grouping, ...

I am rn not sure what would constitute good unit tests. are there worked examples where you manually derived the true solutions? then one could use those here. otherwise we could maybe define a set of useful unit tests.

Reply:
I agree. 

With a particular seed and settings, the parameter inference is consistent with a pretty good optimal plateau for simple models (RAF, Boehm), even though they're not the true parameters (I believe non-importantly for this). I can then write test using optimal outer parameters of these optimizations, for instance:
- check problem is constructed as should be, with correct group dictionaries (inner_problem.groups[group] dictionaries)
- inner optimization results are the ones they should be: surrogate data, lower, upper bounds.
- check fval is the same.
- check that gradient calculated is close to 0 to check solver gradient calculation.

I believe that would be enough/test all features.

Reply:
Added some new tests in the last commit.
Some tests are failing, not sure why, will see tomorrow.